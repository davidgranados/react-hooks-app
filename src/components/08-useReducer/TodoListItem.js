import React from "react";
import classNames from "classnames";

export const TodoListItem = ({ todo, index, onDelete, onToggle }) => {
  const handleDelete = ({ target }) => {
    onDelete(parseInt(target.dataset.id));
  };
  const handleToggle = ({ target }) => {
    onToggle(parseInt(target.dataset.id));
  };
  return (
    <li key={todo.id} className="list-group-item">
      <p
        className={classNames({ complete: todo.done })}
        data-id={todo.id}
        onClick={handleToggle}
      >
        {index}. {todo.desc}
      </p>
      <button
        className="btn btn-danger"
        data-id={todo.id}
        onClick={handleDelete}
      >
        Borrar
      </button>
    </li>
  );
};
