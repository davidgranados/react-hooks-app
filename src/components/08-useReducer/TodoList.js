import React from "react";
import { TodoListItem } from "./TodoListItem";

export const TodoList = ({ todoList, onDelete, onToggle }) => {
  return (
    <ul className="list-group list-group-flush">
      {todoList.map((todo, i) => (
        <TodoListItem
          key={todo.id}
          todo={todo}
          index={i + 1}
          onDelete={onDelete}
          onToggle={onToggle}
        />
      ))}
    </ul>
  );
};
