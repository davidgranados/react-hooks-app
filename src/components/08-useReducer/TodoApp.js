import React, { useReducer, useEffect } from "react";
import { todoReducer } from "./todoReducer";

import { TodoList } from "./TodoList";
import { TodoAdd } from "./TodoAdd";
import "./styles.css";

const init = () => {
  return JSON.parse(localStorage.getItem("todoList")) || [];
};

export const TodoApp = () => {
  const [todoList, dispatch] = useReducer(todoReducer, [], init);

  useEffect(() => {
    localStorage.setItem("todoList", JSON.stringify(todoList));
  }, [todoList]);

  const onDelete = (todoId) => {
    dispatch({
      type: "delete",
      payload: todoId,
    });
  };

  const onToggle = (todoId) => {
    dispatch({
      type: "toggle",
      payload: todoId,
    });
  };

  const onAdd = (newTodo) => {
    dispatch({
      type: "add",
      payload: newTodo,
    });
  };

  return (
    <div>
      <h1>TodoApp ( {todoList.length} ) </h1>
      <hr />

      <div className="row">
        <div className="col-7">
          <TodoList todoList={todoList} onDelete={onDelete} onToggle={onToggle} />
        </div>

        <div className="col-5">
          <TodoAdd onAdd={onAdd} />
        </div>
      </div>
    </div>
  );
};
