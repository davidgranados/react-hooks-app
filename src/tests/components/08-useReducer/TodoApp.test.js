import React from "react";
import { shallow, mount } from "enzyme";
import { TodoApp } from "../../../components/08-useReducer/TodoApp";
import { demoTodo } from "../../fixtures/demoTodo";
import { act } from "@testing-library/react";

describe("Pruebas en <TodoApp />", () => {
  const wrapper = shallow(<TodoApp />);

  Storage.prototype.setItem = jest.fn(() => {});

  it("debe de mostrarse correctamente", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("debe de agregar un TODO", () => {
    const wrapper = mount(<TodoApp />);

    act(() => {
      wrapper.find("TodoAdd").prop("onAdd")(demoTodo[0]);
      wrapper.find("TodoAdd").prop("onAdd")(demoTodo[1]);
    });

    expect(wrapper.find("h1").text().trim()).toBe("TodoApp ( 2 )");
    expect(localStorage.setItem).toHaveBeenCalledTimes(2);
  });

  it("debe de eliminar un todo", () => {
    console.log(wrapper);
    wrapper.find("TodoAdd").prop("onAdd")(demoTodo[0]);
    console.log(wrapper.find("h1").text());
    wrapper.find("TodoList").prop("onDelete")(demoTodo[0].id);

    expect(wrapper.find("h1").text().trim()).toBe("TodoApp ( 0 )");
  });
});
