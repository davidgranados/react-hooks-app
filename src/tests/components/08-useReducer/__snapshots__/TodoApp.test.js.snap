// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <TodoApp /> debe de mostrarse correctamente 1`] = `
<div>
  <h1>
    TodoApp ( 
    0
     ) 
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-7"
    >
      <TodoList
        onDelete={[Function]}
        onToggle={[Function]}
        todoList={Array []}
      />
    </div>
    <div
      className="col-5"
    >
      <TodoAdd
        onAdd={[Function]}
      />
    </div>
  </div>
</div>
`;
